src/AutoMapper/XpressionMapper/Extensions/MapperExtensions.cs

```diff
@@ -235,13 +263,46 @@ private static void AddUnderlyingTypes(this Dictionary<Type, Type> typeMappings,
             if (sourceArguments.Count != destArguments.Count)
                 throw new ArgumentException(Resource.invalidArgumentCount);
 
-            return sourceArguments.Aggregate(typeMappings, (dic, next) =>
+            for (int i = 0; i < sourceArguments.Count; i++)
             {
-                if (!dic.ContainsKey(next) && next != destArguments[sourceArguments.IndexOf(next)])
-                    dic.AddTypeMapping(next, destArguments[sourceArguments.IndexOf(next)]);
-
-                return dic;
-            });
+                if (!typeMappings.ContainsKey(sourceArguments[i]) && sourceArguments[i] != destArguments[i])
+                    typeMappings.AddTypeMapping(configurationProvider, sourceArguments[i], destArguments[i]);
+            }
+
+            return typeMappings;
+        }
+
+        private static void FindChildPropertyTypeMaps(this Dictionary<Type, Type> typeMappings, IConfigurationProvider ConfigurationProvider, Type source, Type dest)
+        {
+            //The destination becomes the source because to map a source expression to a destination expression,
+            //we need the expressions used to create the source from the destination
+            var typeMap = ConfigurationProvider.ResolveTypeMap(sourceType: dest, destinationType: source);
+
+            if (typeMap == null)
+                return;
+
+            FindMaps(typeMap.GetPropertyMaps().ToList());
```

Comment:
I think an explicit return would make things more readable.

Reply:
???? - not clear on this.  Return from FindMaps?